File: hw/toapanta/u3/hw29Singleton/AlarmSystemSingletonJs/controller/AlarmController.js
const Database = require('./Database');
const AlarmService = require('../model/AlarmService');

class AlarmController {
    static async handleUpdateStock(newLimit) {
        const service = await AlarmService.getInstance();
        await service.updateMinStock(newLimit);
        return await this.checkInventory();
    }

    static async checkInventory() {
        const service = await AlarmService.getInstance();
        const db = await Database.getDatabase();
        const products = await db.collection("Products").find().toArray();
        
        return products.filter(product => service.isLowStock(product))
                       .map(product => ({
                           name: product.name, 
                           stock: product.stock, 
                           minLimit: service.getMinStock() 
                       }));
    }

    static async run(window) {
        const alerts = await this.checkInventory();
        if (alerts.length > 0) {
            window.webContents.send('initial-alerts', alerts);
        }
    }
}

module.exports = AlarmController;

File: hw/toapanta/u3/hw29Singleton/AlarmSystemPy/controller/AlarmController.py
from tkinter import messagebox
from model.AlarmService import AlarmService
from controller.Database import Database

class AlarmController:
    def __init__(self):
        self.view = None
        self.service = AlarmService.get_instance()

    def set_view(self, view):
        self.view = view

    def run(self):
        self.check_inventory()
        self.view.mainloop()

    def handle_update(self, value):
        try:
            new_limit = int(value)
            self.service.update_min_stock(new_limit)
            messagebox.showinfo("Success", "Stock updated!")
            self.check_inventory()
        except ValueError:
            messagebox.showerror("Error", "Invalid number")

    def check_inventory(self):
        db = Database.get_database()
        products = db["Products"].find()
        for product in products:
            if self.service.is_low_stock(product["stock"]):
                self.view.show_alert(f"Product: {product['name']}\nStock: {product['stock']}")